typing/typecore.ml

```diff
@@ -3721,15 +3731,28 @@ and type_expect_
         exp_type = newty (Tpackage (p, nl, tl'));
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
-  | Pexp_open (ovf, lid, e) ->
-      let (path, newenv) = !type_open ovf env sexp.pexp_loc lid in
-      let exp = type_expect newenv e ty_expected_explained in
-      { exp with
-        exp_extra = (Texp_open (ovf, path, lid, newenv), loc,
-                     sexp.pexp_attributes) ::
-                      exp.exp_extra;
-      }
-
+  | Pexp_open (ovf, me, e) -> begin
+      match !gen_mod_ident me with
+      | None ->
+        let (_id, tme, newenv) = !type_open ovf env loc me in
+        let exp = type_expect newenv e ty_expected_explained in
+        let od = { open_expr=tme; open_override=ovf;
+                   open_loc=loc; open_env=newenv;
+                   open_attributes=sexp.pexp_attributes } in
+        { exp with
+          exp_extra = (Texp_open od, loc,
+                       sexp.pexp_attributes) ::
+                        exp.exp_extra;
+        }
+      | Some {Ident.name} ->
+        let pmod = Pmod_ident ({txt=Longident.Lident name; loc}) in
+        let popen = Pexp_open (ovf, {me with pmod_desc = pmod}, e) in
+        let pexp_desc = Pexp_letmodule ({txt=name;loc}, me,
+          {sexp with pexp_desc=popen}) in
+        let sexp = {pexp_desc; pexp_loc=loc;
```

Comment:
Couldn't we deal more directly with local open on non-identifier expressions?  The type-checking rule for Pexp_letmodule is not very complex.

Note: in the current version, sexp.pexp_attributes are "duplicated" on the Pexp_open and Pexp_letmodule nodes.